File: packages/swift/Sources/Blasphem/Locales.swift
import Foundation

private let canonical: [String: String] = Dictionary(
    uniqueKeysWithValues: locales.flatMap { entry in
        ([entry.code] + entry.aliases).map { ($0, entry.code) }
    }
)

private let registryOrder: [String: Int] = Dictionary(
    uniqueKeysWithValues: locales.enumerated().map { ($1.code, $0) }
)

/// Lowercases, resolves aliases, rejects unknown codes, and returns registry order without repeats.
func normalizeLocales(_ requested: [String]) throws -> [String] {
    guard !requested.isEmpty else {
        throw BlasphemError(code: .localesEmpty, message: "pass at least one locale, such as [\"en\"]")
    }
    let codes = try requested.map { raw -> String in
        guard let code = canonical[raw.trimmingCharacters(in: .whitespaces).lowercased()] else {
            throw BlasphemError(code: .localeUnsupported, message: "unsupported locale \"\(raw)\"")
        }
        return code
    }
    return Array(Set(codes)).sorted { (registryOrder[$0] ?? 0) < (registryOrder[$1] ?? 0) }
}

File: packages/android/engine/src/main/kotlin/me/sospedra/blasphem/Locales.kt
package me.sospedra.blasphem

private val canonical: Map<String, String> =
    LOCALES.flatMap { (code, aliases) -> (aliases + code).map { it to code } }.toMap()

private val registryOrder: Map<String, Int> =
    LOCALES.withIndex().associate { (index, entry) -> entry.first to index }

/** Lowercases, resolves aliases, rejects unknown codes, and returns registry order without repeats. */
internal fun normalizeLocales(requested: List<String>): List<String> {
    if (requested.isEmpty()) {
        throw BlasphemException(BlasphemException.Code.LOCALES_EMPTY, "pass at least one locale, such as listOf(\"en\")")
    }
    val codes = requested.map { raw ->
        canonical[raw.trim().lowercase()]
            ?: throw BlasphemException(BlasphemException.Code.LOCALE_UNSUPPORTED, "unsupported locale \"$raw\"")
    }
    return codes.distinct().sortedBy { registryOrder.getValue(it) }
}
